{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "13813541-76f2-4bce-a2c0-a52d639004b9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from mygene import MyGeneInfo\n",
    "import pandas as pd\n",
    "DATA_DIR = \"/orcd/pool/003/dbertsim_shared/ukb/\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8a30e7f-0a1d-4e83-b94b-9748990a2ad4",
   "metadata": {},
   "source": [
    "## Protein to BED"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "ab9abf8e-2a5e-4054-893f-165a995b6eb4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "outcome = \"prostate_cancer\"\n",
    "\n",
    "feats = pd.read_csv(f\"features/{outcome}/top300_features_proteomics.csv\")\n",
    "proteins = list(feats['Feature'])\n",
    "proteins = [x.replace('olink_', '') for x in proteins]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "87977829-5f3a-4081-8174-7b61b758b0b6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_39856/2756456322.py:2: DtypeWarning: Columns (3028,3029,3030,3031) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df  = pd.read_csv('/orcd/pool/003/dbertsim_shared/ukb/ukb_cancer_test.csv')\n"
     ]
    }
   ],
   "source": [
    "# write for all proteins\n",
    "df  = pd.read_csv('/orcd/pool/003/dbertsim_shared/ukb/ukb_cancer_test.csv')\n",
    "olink_cols = [c for c in df.columns if \"olink\" in c]\n",
    "proteins = [x.replace('olink_', '') for x in olink_cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "4c6c0221-f220-484d-8a24-7ff1aa217e0b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Input sequence provided is already in string format. No operation performed\n",
      "Input sequence provided is already in string format. No operation performed\n",
      "1 input query terms found dup hits:\t[('siglec5', 2)]\n",
      "45 input query terms found no hit:\t['amy1a_amy1b_amy1c', 'anp32c', 'arntl', 'bap18', 'bola2_bola2b', 'btnl10', 'c7orf50', 'cenpj', 'cer\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genes before selecting for chromosomes: 3480\n",
      "Number of genes after selecting for chromosomes: 34\n",
      "Number of genes before dropping duplicates: 34\n",
      "Number of genes after dropping duplicates: 34\n",
      "BED file written to: features/hugo_genes_GRCh38_ch18.bed\n"
     ]
    }
   ],
   "source": [
    "# -----------------------------\n",
    "# 1. INPUT: HUGO / HGNC symbols\n",
    "# -----------------------------\n",
    "hugo_genes = proteins   # Filter to standard chromosomes if you want\n",
    "valid_chrs = ['18']\n",
    "# valid_chrs = [str(i) for i in range(1, 23)] + [\"X\", \"Y\", \"MT\", \"M\"]\n",
    "\n",
    "mg = MyGeneInfo()\n",
    "\n",
    "# -----------------------------\n",
    "# 2. Query mygene.info\n",
    "#    - scopes='symbol' : we query by gene symbol\n",
    "#    - fields: we ask for genomic coordinates and identifiers\n",
    "# -----------------------------\n",
    "res = mg.querymany(\n",
    "    hugo_genes,\n",
    "    scopes=\"symbol\",\n",
    "    fields=\"symbol,genomic_pos.chr,genomic_pos.start, genomic_pos.end\", # got rid of strand field\n",
    "    species=\"human\"\n",
    ")\n",
    "\n",
    "# Convert to DataFrame for easier manipulation\n",
    "df = pd.DataFrame(res)\n",
    "df = df[df[\"genomic_pos\"].notnull()]\n",
    "\n",
    "# Handle possible list-valued fields (ensembl.gene can be list)\n",
    "def _first(x):\n",
    "    if isinstance(x, dict) and len(x) > 0:\n",
    "        return list(x.values())\n",
    "    return x\n",
    "\n",
    "df[\"genomic_pos\"] = df[\"genomic_pos\"].apply(\n",
    "    lambda x: x if isinstance(x, list) else [x]\n",
    ")\n",
    "df = df.explode(\"genomic_pos\").reset_index(drop=True)\n",
    "pos_df = pd.json_normalize(df['genomic_pos'])\n",
    "df = pd.concat([df.drop(columns=[\"genomic_pos\"]), pos_df], axis=1)\n",
    "\n",
    "print(f\"Number of genes before selecting for chromosomes: {len(df)}\")\n",
    "df = df[df[\"chr\"].isin(valid_chrs)]\n",
    "print(f\"Number of genes after selecting for chromosomes: {len(df)}\")\n",
    "\n",
    "# -----------------------------\n",
    "# 3. Build BED DataFrame (0-based start)\n",
    "# -----------------------------\n",
    "bed = pd.DataFrame({\n",
    "    \"chr\": \"chr\" + df[\"chr\"].astype(str),\n",
    "    \"start\": df[\"start\"].astype(int) - 1,  # 0-based\n",
    "    \"end\": df[\"end\"].astype(int),\n",
    "    \"name\": df[\"symbol\"],\n",
    "    # \"score\": 0,\n",
    "    # \"strand\": df[\"strand\"]\n",
    "})\n",
    "\n",
    "# Optional: deduplicate per gene symbol\n",
    "print(f\"Number of genes before dropping duplicates: {len(bed)}\")\n",
    "bed = bed.drop_duplicates()\n",
    "print(f\"Number of genes after dropping duplicates: {len(bed)}\")\n",
    "\n",
    "\n",
    "# -----------------------------\n",
    "# 4. Write BED file\n",
    "# -----------------------------\n",
    "# out_file = f\"hugo_genes_GRCh38_genes_{outcome}.bed\"\n",
    "out_file = f\"features/hugo_genes_GRCh38_ch18.bed\"\n",
    "bed.to_csv(out_file, sep=\"\\t\", header=False, index=False)\n",
    "\n",
    "print(\"BED file written to:\", out_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "3cf71e4f-38e4-4022-8581-d529321aa935",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "chr1     330\n",
       "chr19    230\n",
       "chr2     197\n",
       "chr11    178\n",
       "chr17    168\n",
       "chr12    152\n",
       "chr6     149\n",
       "chr3     135\n",
       "chr7     128\n",
       "chr4     127\n",
       "chr5     127\n",
       "chr9     119\n",
       "chr16    115\n",
       "chr10    110\n",
       "chr8     102\n",
       "chr14     91\n",
       "chrX      88\n",
       "chr20     77\n",
       "chr22     74\n",
       "chr15     73\n",
       "chr13     47\n",
       "chr18     34\n",
       "chr21     29\n",
       "chrY       3\n",
       "Name: chr, dtype: int64"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bed['chr'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "65bb3916-07b6-46d3-a0d3-0eb41fd508d2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "34"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(bed.loc[bed['chr'] == 'chr18'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "431170e4-21df-40c7-adb3-a8b87bff14fc",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>chr18</td>\n",
       "      <td>31376776</td>\n",
       "      <td>31414912</td>\n",
       "      <td>DSG4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>chr18</td>\n",
       "      <td>45983535</td>\n",
       "      <td>46072296</td>\n",
       "      <td>PSTPIP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227</th>\n",
       "      <td>chr18</td>\n",
       "      <td>316736</td>\n",
       "      <td>500722</td>\n",
       "      <td>COLEC12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>chr18</td>\n",
       "      <td>32185068</td>\n",
       "      <td>32220404</td>\n",
       "      <td>MEP1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>327</th>\n",
       "      <td>chr18</td>\n",
       "      <td>74534399</td>\n",
       "      <td>74587212</td>\n",
       "      <td>CNDP1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr     start       end     name\n",
       "145  chr18  31376776  31414912     DSG4\n",
       "181  chr18  45983535  46072296  PSTPIP2\n",
       "227  chr18    316736    500722  COLEC12\n",
       "240  chr18  32185068  32220404    MEP1B\n",
       "327  chr18  74534399  74587212    CNDP1"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bed # prostate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "eaffaa63-0438-4f63-8897-4eee12af9a95",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr18</td>\n",
       "      <td>721587</td>\n",
       "      <td>813276</td>\n",
       "      <td>YES1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>chr18</td>\n",
       "      <td>32185068</td>\n",
       "      <td>32220404</td>\n",
       "      <td>MEP1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>chr18</td>\n",
       "      <td>63970028</td>\n",
       "      <td>64019779</td>\n",
       "      <td>SERPINB8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr     start       end      name\n",
       "29   chr18    721587    813276      YES1\n",
       "120  chr18  32185068  32220404     MEP1B\n",
       "162  chr18  63970028  64019779  SERPINB8"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bed # breast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "30c67d66-75ea-4336-877e-5aa56da3cddd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>chr18</td>\n",
       "      <td>31497623</td>\n",
       "      <td>31551351</td>\n",
       "      <td>DSG2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>chr18</td>\n",
       "      <td>52340196</td>\n",
       "      <td>53535899</td>\n",
       "      <td>DCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>332</th>\n",
       "      <td>chr18</td>\n",
       "      <td>7566781</td>\n",
       "      <td>8406861</td>\n",
       "      <td>PTPRM</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr     start       end   name\n",
       "156  chr18  31497623  31551351   DSG2\n",
       "243  chr18  52340196  53535899    DCC\n",
       "332  chr18   7566781   8406861  PTPRM"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bed # lung"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d04a2fe-9c8e-48d1-b6ec-6d32bca54102",
   "metadata": {},
   "source": [
    "## Intersect Protein and SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "eedc5537-c11a-4386-8623-6bf2a6f86aa3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# run locally to get intersection\n",
    "\n",
    "# bedtools intersect -a SNPs.bed -b protein.bed -wa -wb | awk 'BEGIN{OFS=\"\\t\"}{print $1,$2,$3,$4,$8}' > SNP_gene_overlaps.bed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3de1a5ec-7fd3-42fa-8afd-098f43faac1f",
   "metadata": {},
   "source": [
    "## SNPs to RSID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "a24616b7-8216-46f8-a4a3-efdf8c866bd0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "outcome = \"prostate\"\n",
    "\n",
    "df = pd.read_csv(f\"bed/SNP_gene_overlaps_{outcome}_cancer.bed\", sep=\"\\t\", header=None, \\\n",
    "                names=[\"chr\", \"start\", \"end\", \"alleles\", \"gene\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "4fdd43d3-4580-4467-9e96-0e2e6df6a928",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Split C>T into ref and alt\n",
    "df[[\"ref\", \"alt\"]] = df[\"alleles\"].str.split(\">\", expand=True)\n",
    "\n",
    "# Remove \"chr\" and use 1-based position (end) for the ID\n",
    "df[\"rsid\"] = (\n",
    "    df[\"chr\"].str.replace(\"chr\", \"\", regex=False)\n",
    "    + \":\" + df[\"end\"].astype(str)\n",
    "    + \":\" + df[\"ref\"]\n",
    "    + \":\" + df[\"alt\"]\n",
    ")\n",
    "\n",
    "variant_map = pd.read_csv(f\"{DATA_DIR}/bgen/ch18/c18_b0_v1_variants.csv\")\n",
    "variant_map_filtered = variant_map.loc[variant_map['rsid'].isin(df['rsid'])]\n",
    "variant_map_filtered.to_csv(f\"bed/variants_overlaps_{outcome}_cancer.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "7f24516e-127f-4002-919e-6f190c5364c3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>rsid</th>\n",
       "      <th>allele_ids</th>\n",
       "      <th>variant_idx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3292</th>\n",
       "      <td>18</td>\n",
       "      <td>319951</td>\n",
       "      <td>18:319951:T:G</td>\n",
       "      <td>T,G</td>\n",
       "      <td>3292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3293</th>\n",
       "      <td>18</td>\n",
       "      <td>319956</td>\n",
       "      <td>18:319956:C:T</td>\n",
       "      <td>C,T</td>\n",
       "      <td>3293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3294</th>\n",
       "      <td>18</td>\n",
       "      <td>319970</td>\n",
       "      <td>18:319970:G:A</td>\n",
       "      <td>G,A</td>\n",
       "      <td>3294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3295</th>\n",
       "      <td>18</td>\n",
       "      <td>319970</td>\n",
       "      <td>18:319970:G:T</td>\n",
       "      <td>G,T</td>\n",
       "      <td>3295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3296</th>\n",
       "      <td>18</td>\n",
       "      <td>319973</td>\n",
       "      <td>18:319973:G:A</td>\n",
       "      <td>G,A</td>\n",
       "      <td>3296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395762</th>\n",
       "      <td>18</td>\n",
       "      <td>74584653</td>\n",
       "      <td>18:74584653:T:C</td>\n",
       "      <td>T,C</td>\n",
       "      <td>395762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395763</th>\n",
       "      <td>18</td>\n",
       "      <td>74584655</td>\n",
       "      <td>18:74584655:T:G</td>\n",
       "      <td>T,G</td>\n",
       "      <td>395763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395764</th>\n",
       "      <td>18</td>\n",
       "      <td>74584656</td>\n",
       "      <td>18:74584656:G:A</td>\n",
       "      <td>G,A</td>\n",
       "      <td>395764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395765</th>\n",
       "      <td>18</td>\n",
       "      <td>74584658</td>\n",
       "      <td>18:74584658:G:A</td>\n",
       "      <td>G,A</td>\n",
       "      <td>395765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395766</th>\n",
       "      <td>18</td>\n",
       "      <td>74584660</td>\n",
       "      <td>18:74584660:C:A</td>\n",
       "      <td>C,A</td>\n",
       "      <td>395766</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7327 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom       pos             rsid allele_ids  variant_idx\n",
       "3292       18    319951    18:319951:T:G        T,G         3292\n",
       "3293       18    319956    18:319956:C:T        C,T         3293\n",
       "3294       18    319970    18:319970:G:A        G,A         3294\n",
       "3295       18    319970    18:319970:G:T        G,T         3295\n",
       "3296       18    319973    18:319973:G:A        G,A         3296\n",
       "...       ...       ...              ...        ...          ...\n",
       "395762     18  74584653  18:74584653:T:C        T,C       395762\n",
       "395763     18  74584655  18:74584655:T:G        T,G       395763\n",
       "395764     18  74584656  18:74584656:G:A        G,A       395764\n",
       "395765     18  74584658  18:74584658:G:A        G,A       395765\n",
       "395766     18  74584660  18:74584660:C:A        C,A       395766\n",
       "\n",
       "[7327 rows x 5 columns]"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_map_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33364c86-866b-433b-936a-e1f78c63360b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
